torchtune/utils/_distributed.py

```diff
@@ -212,3 +214,59 @@ def lora_wrap_fsdp(module: nn.Module, recurse: bool, **kwargs):
         return isinstance(module, tuple(modules_to_wrap))
 
     return lora_wrap_fsdp
+
+
+def get_full_finetune_fsdp_wrap_policy(
+    memory_efficient_wrapping: bool, modules_to_wrap: Set[Type]
+) -> FSDPPolicyType:
+    """
+    Retrieves an FSDP wrapping policy based on the specified ``model_type`` and ``modules_to_wrap``.
+    Specifically, for Llama3 model variants, the policy returned will wrap the model's token embedding
+    and output projection in addition to the modules specified in ``modules_to_wrap`` to maximize memory savings.
+    For all other models, only the modules specified will be wrapped.
+    Args:
+        memory_efficient_wrapping (bool): If ``True``, will also wrap embedding and output projection layers
+        with FSDP.
+        modules_to_wrap (Set[Type]): Set of module types to wrap.
+
+    Note:
+        ``memory_efficient_wrapping`` memory improvements have currently only been verified on llama3 workloads,
+            to provide ~15% memory improvement (when used alongside AC memory efficient wrapping). Other workloads
+            have not been verified and may not see the same improvements.
+    Returns:
+        FSDPPolicyType: Wrapping policy that can be passed into ``FullyShardedDataParallel``.
+    """
+    if memory_efficient_wrapping:
+        return _llama3_full_fsdp_wrap_policy(modules_to_wrap=modules_to_wrap)
+    else:
+        return ModuleWrapPolicy(modules_to_wrap)
+
+
+def _llama3_full_fsdp_wrap_policy(modules_to_wrap: Set[Type]) -> FSDPPolicyType:
+    """
+    A default policy for wrapping Llama-3 style models for full finetuning using FSDP. Specifically,
+    this will wrap the model's token embedding and output projection into their own FSDP units to
+    maximize memory savings. After this is done, model will also be hierarchically wrapped
+    based on nn.Module types specified in ``modules_to_wrap``. This function assumes that the
+    input model has an attribute ``output`` that is a nn.Linear which is the model's output projection.
+    Args:
+        modules_to_wrap (Set[Type]): nn.Module types to recursively wrap
+    Returns:
+        FSDPPolicyType: Wrapping policy that can be passed into ``FullyShardedDataParallel``.
+    """
+    modules_to_wrap.add(torch.nn.Embedding)
+
+    def llama3_wrap(module: nn.Module, recurse: bool, **kwargs):
+        # Label that output_proj should be wrapped individually.
+        if isinstance(module, modules.TransformerDecoder):
+            module.output._wrap = True
```

Comment:
@ebsmothers how does this impact some of the transformer decoder changes that we're planning to do?

Reply:
What does ```_wrap``` refer to here? I tried to look at the FSDP docs and the nn.Module docs and it's not immediately clear to me what this is doing? Sorry if I'm being dense on this.

Reply:
Re the transformer decoder changes, I think the main thing is that the `if isinstance` check may need to be generalized (since now e.g. our `TransformerDecoderLM` or `TransformerDecoderClassifier` classes will both have output layers). But realistically I think the main use case will still be for when we're projecting to vocab_size, so maybe just directly replacing with `TransformerDecoderLM` (or whatever we're calling it) will be sufficient here.

Reply:
@ebsmothers IIUC, it doesn't seem that we need any particular changes right now since the classifier support is only there for mistral at the moment? Although yeah we should keep an eye in the future - if we send a classifier model to this wrapping policy, it might not end up wrapping things correctly. Probably can consider it in a future PR?